Adding regex check for crossfire gps format when reading logs in tele…

Harpalyke opened the pull request:
…metry simulator

Adding a regex check when the telemetry simulator parses gps cords from a log file to see if the if the cords are in the crossfire gps log format. 
The GPS telemetry sensor from crossfire logs the coordinates as latitude with a precision of six decimal places followed by a space and then longitude also to six decimal places. e.g. "53.279969 -7.701416"
When the cords are already in this format their is no need to call convertDegMin on the values

Addresses issue #5346

Harpalyke committed:
Adding regex check for crossfire gps format when reading logs in telemetry simulator

Adding a regex check when the telemetry simulator parses gps cords from a log file to see if the if the cords are in the crossfire gps log format. 
The GPS telemetry sensor from crossfire logs the coordinates as latitude with a precision of six decimal places followed by a space and then longitude also to six decimal places. e.g. "53.279969 -7.701416"
When the cords are already in this format their is no need to call convertDegMin on the values

bsongis commented on companion/src/simulation/telemetrysimu.cpp:
```diff
@@ -868,8 +868,19 @@ QString TelemetrySimulator::LogPlaybackController::convertGPS(QString input)
+  QRegularExpression crossfireCoordinateFormatRegex("^([-+]?\\d{1,2}[.]\\d{6})$");
+  QRegularExpressionMatch latFormatMatch = crossfireCoordinateFormatRegex.match(lonLat[0]);
+  QRegularExpressionMatch lonFormatMatch = crossfireCoordinateFormatRegex.match(lonLat[1]);
+  if(lonFormatMatch.hasMatch() && latFormatMatch.hasMatch()){
```
space between `if` and `(`

bsongis commented on companion/src/simulation/telemetrysimu.cpp:
```diff
@@ -868,8 +868,19 @@ QString TelemetrySimulator::LogPlaybackController::convertGPS(QString input)
+  if(lonFormatMatch.hasMatch() && latFormatMatch.hasMatch()){
+	  lat = lonLat[0].toDouble();
+	  lon = lonLat[1].toDouble();
+  }else{
```
newline before `else` and space between `else` and `{`

bsongis reviewed (commented):
just cosmetics before I press the merge button so that it is in next release

Harpalyke committed:
Fixing code styling

Harpalyke commented:
Updated. Thanks!

bsongis commented on companion/src/simulation/telemetrysimu.cpp:
```diff
@@ -874,10 +874,12 @@ QString TelemetrySimulator::LogPlaybackController::convertGPS(QString input)
 	  lon = lonLat[1].toDouble();
-  }else{
+  }
+  else
```
no newline after the else! sorry to be so picky!

Harpalyke committed:
Updating code style

Harpalyke commented:
No problem. I just misread it, I should have checked one of the other else stmts in the code anyway.

bsongis commented on companion/src/simulation/telemetrysimu.cpp:
```diff
@@ -868,8 +868,20 @@ QString TelemetrySimulator::LogPlaybackController::convertGPS(QString input)
+  QRegularExpressionMatch latFormatMatch = crossfireCoordinateFormatRegex.match(lonLat[0]);
+  QRegularExpressionMatch lonFormatMatch = crossfireCoordinateFormatRegex.match(lonLat[1]);
+  if (lonFormatMatch.hasMatch() && latFormatMatch.hasMatch()){
+	  lat = lonLat[0].toDouble();
```
and now indentation is bad, we use 2 spaces in this project

bsongis commented on companion/src/simulation/telemetrysimu.cpp:
```diff
@@ -868,8 +868,20 @@ QString TelemetrySimulator::LogPlaybackController::convertGPS(QString input)
+  QRegularExpression crossfireCoordinateFormatRegex("^([-+]?\\d{1,2}[.]\\d{6})$");
+  QRegularExpressionMatch latFormatMatch = crossfireCoordinateFormatRegex.match(lonLat[0]);
+  QRegularExpressionMatch lonFormatMatch = crossfireCoordinateFormatRegex.match(lonLat[1]);
+  if (lonFormatMatch.hasMatch() && latFormatMatch.hasMatch()){
```
and a space before the `{`

bsongis reviewed (commented):
cosmetics again

Harpalyke committed:
Updating code style

mpaperno commented:
The regex check for every line of the file will be a fairly expensive operation.  At the least I'd recommend only defining that regex once (as a class member or as static in just the parsing function) -- this will prevent the need for Qt to "compile" the expression on every execution.  

Another option, assuming all GPS data is formatted the same way in every line of a given log file, is to determine the format on the first line of the log and then use that determination for the rest of the lines -- so the regex needs to be executed only once per file.

Lastly, if the log file GPS format itself could be normalized (always log in same format), that may be a better option, but I haven't looked at the logging routines at all so maybe that is less practical in this case.

mpaperno commented on companion/src/simulation/telemetrysimu.h:
```diff
@@ -27,6 +27,7 @@
 #include <QDateTime>
 #include <QtCore/qmath.h>
 #include <QFileDialog>
+#include <QRegularExpression>
```
Any particular reason to put this in the header file?  Since there's no QRegularExpression declared in the header, this would probably make more sense in the cpp file.  Not a big deal either way.

mpaperno commented on companion/src/simulation/telemetrysimu.cpp:
```diff
@@ -868,8 +868,20 @@ QString TelemetrySimulator::LogPlaybackController::convertGPS(QString input)
+  double lon, lat;
+  // crossfire GPS sensor logs cords in decimal format with a precision of 6 places
+  // if this format is met there is no need to call convertDegMin
+  QRegularExpression crossfireCoordinateFormatRegex("^([-+]?\\d{1,2}[.]\\d{6})$");
```
Since the captured result is never used, the parenthesis (after `^` and before `$`) could be excluded for a slight optimization of the regex.  Also, if you haven't yet, I'd make sure that a comma is never being used here as a decimal separator (e.g. in EU locales).

Harpalyke committed:
Checking the gps coordinates format once before the full file is processed instead of evaluating the regex for every line of the file.

Optimize regex by removing surrounding parenthesis

Harpalyke commented:
@mpaperno, good point on executing the regex for every line of the file. I've updated the code so it samples the first line as suggested and tidied up the other bits you suggested. 

Also, the data I had been using for testing was generated with a crossfire module in an X9D. I had assumed that it was because of the fact I was using the crossfire that the GPS data had been logged in an incompatible format. However, I've run a quick test logging the GPS data when connecting via an FrSky XSR and it appears that the gps format being logged is the same as with the crossfire instead of the expected "DDmm.mmmmH DDDmm.mmmmH" format noted in the code. 
Maybe the logging format was changed on the radio side but the expected format on the telemetry simulator wasn't updated to match?

bsongis commented:
thanks! still not 100% compliant with our coding rules, but not so far.

Harpalyke commented:
@bsongis, OK, sorry about that, I'll make sure to follow the style guide more closely next time. Thanks!
